Update U2FServerReferenceImpl for Transfer of Access

alextaka opened the pull request:
Update U2FServerReferenceImpl.java to allow for processing of TransferAccessResponses and AuthenticationResponses as SignResponses.
Added basic tests.

alextaka committed:
Update U2FServerReferenceImpl for Transfer of Access

Update U2FServerReferenceImpl.java to allow for processing of TransferAccessResponses and AuthenticationResponses as SignResponses.
Added basic tests.

alextaka commented:
@leshi PTAL

leshi commented on u2f-ref-code/java/src/com/google/u2f/server/impl/U2FServerReferenceImpl.java:
```diff
@@ -358,4 +534,23 @@ static String canonicalizeOrigin(String url) {
+      byte controlFlags = inputStream.readByte();
+
+      return ControlFlags.fromByte(controlFlags);
+
```
remove extra new line

alextaka replied:
done

leshi commented on u2f-ref-code/java/src/com/google/u2f/server/impl/U2FServerReferenceImpl.java:
```diff
@@ -237,7 +243,7 @@ public SecurityKeyData processSignResponse(SignResponse signResponse, long curre
     String browserData = new String(Base64.decodeBase64(browserDataBase64));
     byte[] rawSignData = Base64.decodeBase64(rawSignDataBase64);
-
+    
```
remove spaces at end of line

leshi replied:
http://stackoverflow.com/questions/4438306/how-to-remove-trailing-whitespaces-with-sed

alextaka replied:
done

leshi commented on u2f-ref-code/java/src/com/google/u2f/server/impl/U2FServerReferenceImpl.java:
```diff
@@ -246,9 +252,176 @@ public SecurityKeyData processSignResponse(SignResponse signResponse, long curre
+    return newSecurityKeyData;
+  }
+
+  private void checkTransferAccessMessageChain(TransferAccessMessage[] transferAccessMessages,
```
this method could use comments in each section of the checks you make and a summary comment for the whole method

alextaka replied:
done

leshi commented on u2f-ref-code/java/src/com/google/u2f/server/impl/U2FServerReferenceImpl.java:
```diff
@@ -246,9 +252,176 @@ public SecurityKeyData processSignResponse(SignResponse signResponse, long curre
+
+    for (TransferAccessMessage transferAccessMessage : transferAccessMessages) {
+      Log.info("  sequenceNumber:  "
+          + Integer.toString(transferAccessMessage.getMessageSequenceNumber()));
```
do you really need to do explicit type conversion?  Does it not work if you concatenate a String with an Integer?

leshi replied:
also applies to similar usage below

alextaka replied:
done.

leshi commented on u2f-ref-code/java/src/com/google/u2f/server/impl/U2FServerReferenceImpl.java:
```diff
@@ -246,9 +252,176 @@ public SecurityKeyData processSignResponse(SignResponse signResponse, long curre
+    Log.info("  -- Parsed transferAccessMessages -- ");
+
+    for (TransferAccessMessage transferAccessMessage : transferAccessMessages) {
+      Log.info("  sequenceNumber:  "
```
where do you check that the sequence numbers increment/decrement correctly?

alextaka replied:
I check that when I build the Transfer Access Message chain transferAccessMessages[]. The check isn't done here because the sequence numbers were inserted for parsing, which is done in a previous step.

I don't think code to do this explicitly is necessary because:
1. They are included in the signatures
2. The signatures need to be in the correct order in order to have this function complete successfully.

leshi commented on u2f-ref-code/java/src/com/google/u2f/server/impl/U2FServerReferenceImpl.java:
```diff
@@ -358,4 +534,23 @@ static String canonicalizeOrigin(String url) {
   }
+
+  /**
+   * If future versions add control bits, their functionality should be added here. The 7th bit is
```
Start with a description of what this does (last sentence) and then end with what it might do (first sentence)

alextaka replied:
done

leshi commented on u2f-ref-code/java/tests/com/google/u2f/server/impl/U2FServerReferenceImplTest.java:
```diff
@@ -236,6 +237,123 @@ public void testProcessSignResponse_badOrigin() throws U2FException {
+      assertTrue(e.getMessage().contains("bad browserdata: bad type"));
+    }
+  }  
+
```
extra newline

alextaka replied:
done

leshi commented on u2f-ref-code/java/tests/com/google/u2f/server/impl/U2FServerReferenceImplTest.java:
```diff
@@ -236,6 +237,123 @@ public void testProcessSignResponse_badOrigin() throws U2FException {
+      assertTrue(e.getMessage().contains("is not a recognized home origin"));
+    }
+  }  
+
```
what about wrong sequence numbers

alextaka replied:
I didn't explicitly test for this because a wrong sequence number would be caught by a signature that doesn't check out.

For example, say the sequence numbers were (somehow) out of order, but the signatures still all checked out. Is this a problem? There is still a clear transition of trust in the attestation and authentication keys in the correct order, they've just somehow signed them with sequence numbers that aren't in order.

This would not work in our parser, but for the purposes of checking here, I don't think it is important.

alextaka replied:
@leshi I actually already have a check in the parser (rawCodec) and a test in the raw codec test for out of order sequence numbers. It currently does not allow them, which I think is fine.

The question is whether we should add a check to the Server Implementation and tests

leshi replied:
no test needed as per discussion on hangouts

alextaka commented on u2f-ref-code/java/src/com/google/u2f/server/impl/U2FServerReferenceImpl.java:
```diff
@@ -237,7 +243,7 @@ public SecurityKeyData processSignResponse(SignResponse signResponse, long curre
     String browserData = new String(Base64.decodeBase64(browserDataBase64));
     byte[] rawSignData = Base64.decodeBase64(rawSignDataBase64);
-
+    
```
done

alextaka committed:
Addressed first round of comments. Two have been left unchanged pending discussion about checking sequence number order

alextaka committed:
removed another newline

alextaka committed:
replaced comments with log statements in keeping with style
Also discussed with Alexei, we are simply replicating the test from rawCodec in the U2FServerReferenceImplTest to test for out-of-order sequence numbers.

alextaka committed:
replaced comments with log statements in keeping with style

alextaka commented:
@leshi ok, PTAL

All tests pass
Test added for server and out of order sequence numbers
